File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/java/com/divideconquer/algorithms/Quickselect.java
package com.divideconquer.algorithms;

import java.util.Arrays;

public final class Quickselect {
    private Quickselect() {
    }

    public static int kthSmallest(int[] values, int k) {
        if (values == null || values.length == 0) {
            throw new IllegalArgumentException("values must not be empty");
        }
        if (k < 0 || k >= values.length) {
            throw new IllegalArgumentException("k must be inside the array bounds");
        }

        int[] copy = Arrays.copyOf(values, values.length);
        int left = 0;
        int right = copy.length - 1;

        while (left <= right) {
            int pivotIndex = partition(copy, left, right);
            if (pivotIndex == k) {
                return copy[pivotIndex];
            }
            if (pivotIndex < k) {
                left = pivotIndex + 1;
            } else {
                right = pivotIndex - 1;
            }
        }
        throw new IllegalStateException("selection failed");
    }

    private static int partition(int[] values, int left, int right) {
        int pivot = values[right];
        int store = left;
        for (int i = left; i < right; i++) {
            if (values[i] <= pivot) {
                swap(values, store, i);
                store++;
            }
        }
        swap(values, store, right);
        return store;
    }

    private static void swap(int[] values, int i, int j) {
        int temp = values[i];
        values[i] = values[j];
        values[j] = temp;
    }
}

File: 29Items-Projects/7Algorithms/5-DIVIDE-CONQUER/src/main/python/divide_conquer/algorithms/quickselect.py
"""Average-case linear-time selection using partitioning."""


def kth_smallest(values: list[int], k: int) -> int:
    """Return the zero-based kth smallest value without mutating the caller's list."""
    if not values:
        raise ValueError("values must not be empty")
    if k < 0 or k >= len(values):
        raise ValueError("k must be inside the list bounds")

    items = list(values)
    left = 0
    right = len(items) - 1

    while left <= right:
        pivot_index = _partition(items, left, right)
        if pivot_index == k:
            return items[pivot_index]
        if pivot_index < k:
            left = pivot_index + 1
        else:
            right = pivot_index - 1

    raise RuntimeError("selection failed")


def _partition(values: list[int], left: int, right: int) -> int:
    pivot = values[right]
    store = left
    for index in range(left, right):
        if values[index] <= pivot:
            values[store], values[index] = values[index], values[store]
            store += 1
    values[store], values[right] = values[right], values[store]
    return store

